Support VK_EXT_depth_clip_control

ShabbyX opened the pull request:
## Description

This extension adds a viewport pipeline state: negativeOneToOne

ShabbyX committed:
Support VK_EXT_depth_clip_control

This extension adds a viewport pipeline state: negativeOneToOne

baldurk reviewed (changes requested):
This looks great, thank you for implementing this.

I'd ask that you also propagate this through to the UI. Unless UI is onerous to implement for an extension like this it's best to do it at the same time, otherwise it's likely to be forgotten unless someone notices.

In code this could probably be passed as a bool next to `discardRectanglesExclusive` in `VKPipe::ViewState`. In the UI you could either add another boolean tick/cross display, though that's getting a bit full, or alternatively add it as a new column in the viewport list. Although it's not a per-viewport state, I think it's fine to have an extra column saying the NDC depth range is either `0->1` or `-1->1` since a single viewport is the most common case by far anyway, and it's more logically grouped.

If you think there's a better way to display it I'm open to suggestions too.

ShabbyX commented:
> This looks great, thank you for implementing this.
> 
> I'd ask that you also propagate this through to the UI. Unless UI is onerous to implement for an extension like this it's best to do it at the same time, otherwise it's likely to be forgotten unless someone notices.
> 
> In code this could probably be passed as a bool next to `discardRectanglesExclusive` in `VKPipe::ViewState`. In the UI you could either add another boolean tick/cross display, though that's getting a bit full, or alternatively add it as a new column in the viewport list. Although it's not a per-viewport state, I think it's fine to have an extra column saying the NDC depth range is either `0->1` or `-1->1` since a single viewport is the most common case by far anyway, and it's more logically grouped.
> 
> If you think there's a better way to display it I'm open to suggestions too.

Ok I'll see what looks best. It would definitely help _me_ to have this in the UI, but FYI this is a "glemulation" extension. Outside the GL layers, there should be no users. Would you still like to see this hooked up in the UI?

baldurk commented:
My view is that if the extension is implemented it should be implemented to a reasonable standard at minimum including the ability to introspect and have access to the appropriate debugging tools (which in this case is just seeing if it's on or not).

Arguably an extension having insufficient demand like that could be an argument for not supporting it at all upstream (if it's too much work and the value isn't justified), but I don't want to have extensions that are only half-there. Especially for something like this where if someone sets it incorrectly/inadvertently they may have a very hard to debug issue if they can't figure out why the rendering is behaving as it is.

In an ideal world extensions could be tightly gated and only used by specific users or tools, but in practice you can see from #2606 that there are people asking or it in a porting context not just for emulation. And in the past I've seen brand new applications that are written using emulation-only/deprecated functionality. If it's in the API someone will use it, even if it's not widespread, and if it's supported by RenderDoc it should work as well as anything else.

ShabbyX commented:
Understood, I'll take care of it within the next few days.

ShabbyX committed:
Expose Vulkan depth clip range in the UI

This change exposes the
VkPipelineViewportDepthClipControlCreateInfoEXT::negativeOneToOne
pipeline state in the UI.  While not a per-viewport state, this state is
output alongside viewport information.

If the extension is not used, or if this flag is false, [0, 1] is shown
as the depth clip range.  If the flag is true, [-1, 1] is shown.

ShabbyX commented:
> I think it's fine to have an extra column saying the NDC depth range is either 0->1 or -1->1 since a single viewport is the most common case by far anyway, and it's more logically grouped.
> 
> If you think there's a better way to display it I'm open to suggestions too.

No, that seemed to be the nicest way, thanks for the suggestion. I implemented this with the following behavior:

- If the app doesn't use VK_EXT_depth_clip_control, the column is hidden. I did this to avoid cluttering the UI for most apps that don't use this extension
- Otherwise, depending on the clip range, the column shows either `[0, 1]` or `[-1, 1]` (as the clip "range").

I've verified the three possible scenarios both with the UI and the exported HTML.

baldurk commented on renderdoc/api/replay/replay_enums.h:
```diff
@@ -4716,6 +4716,30 @@ enum class AndroidFlags : uint32_t
+  Enabled,
+};
+
+DECLARE_REFLECTION_ENUM(FeatureStatus);
```
This is a nice idea in principle but I think it's overkill for a feature like this, and I also want to avoid as much as possible having cap bits like this which the UI and other consumers using the python API then need to check. Instead any optional feature/functionality can simply be in the disabled/default state that exists when it's not used - whether it's not used or supported by the capture and driver or it's just disabled at the current event.

Also rather than making a separate review comment to be clear - I think it's fine to just have this state always have its column visible. It's not a huge invasive new section and there are several other features which are as-or-more obscure which are always visible e.g. line stipple. Viewport state is not something people look at often anyway, and the dimensions/offsets are still easy to see even with an extra column. If it were a huge imposition with lots of state then arguably it could be displayed conditionally like stream-out is on the geometry shader tab, but in that case there's not much value in distinguishing between not available and disabled.

ShabbyX replied:
That's fair, done.

I also changed "Depth Clip Range" to "NDC Depth Range" to avoid any confusion with other depth-clip-related Vulkan extensions
